// rxjs/_esm2015/internal-compatibility/index.js
export { config } from '../internal/config';
export { InnerSubscriber } from '../internal/InnerSubscriber';
export { OuterSubscriber } from '../internal/OuterSubscriber';
export { Scheduler } from '../internal/Scheduler';
export { AnonymousSubject } from '../internal/Subject';
export { SubjectSubscription } from '../internal/SubjectSubscription';
export { Subscriber } from '../internal/Subscriber';
export { fromPromise } from '../internal/observable/fromPromise';
export { fromIterable } from '../internal/observable/fromIterable';
export { ajax } from '../internal/observable/dom/ajax';
export { webSocket } from '../internal/observable/dom/webSocket';
export { ajaxGet, ajaxPost, ajaxDelete, ajaxPut, ajaxPatch, ajaxGetJSON, AjaxObservable, AjaxSubscriber, AjaxResponse, AjaxError, AjaxTimeoutError } from '../internal/observable/dom/AjaxObservable';
export { WebSocketSubject } from '../internal/observable/dom/WebSocketSubject';
export { CombineLatestOperator } from '../internal/observable/combineLatest';
export { dispatch } from '../internal/observable/range';
export { SubscribeOnObservable } from '../internal/observable/SubscribeOnObservable';
export { Timestamp } from '../internal/operators/timestamp';
export { TimeInterval } from '../internal/operators/timeInterval';
export { GroupedObservable } from '../internal/operators/groupBy';
export { defaultThrottleConfig } from '../internal/operators/throttle';
export { rxSubscriber } from '../internal/
